강원대 FE 안성현 - 2단계 주문하기 페이지 구현하기

anseonghyeon opened the pull request:
1. 로그인 정보를 어디서든 사용할 수 있게 해요.
2. 새로고침을 하더라도 로그인 정보가 유지되도록 해요.
3. 로그인 한 유저는 로그인 버튼 클릭 시 마이페이지(/my)로 연결돼요.
4. 마이페이지는 로그인 한 사람만 접근 가능해요. 로그인 하지 않은 사람이 접속하면 로그인 페이지로 연결해요.
5. 마이페이지에서 로그아웃을 할 수 있어요. 로그아웃 시 로그인(/login) 페이지로 연결돼요.
6. 상품 아이템을 클릭하면 주문하기 페이지로 연결돼요.
7. 만약 로그인이 되어있지 않다면 로그인을 유도하고, 로그인 성공 시 주문하기 페이지로 접근 가능하게 해요.
8. 버튼 클릭 시 아래 유효성 검사 조건이 충족되지 않으면 각 폼 필드에 안내 메시지를 보여줘요.
9. 메시지는 반드시 입력 되어야 해요.
10. 보내는 사람 이름이 반드시 입력 되어야 해요.
11. 받는 사람 이름이 반드시 입력 되어야 해요.
12. 받는사람 전화번호가 반드시 입력되고 전화번호 규칙에 맞아야 해요. (01012341234)
13. 수량은 1개 이상이어야 해요.

anseonghyeon committed:
feat(auth): 로그인 정보를 어디서든 사용할 수 있게 context api에 저장함

anseonghyeon committed:
feat(auth): 세션 스토리지를 사용해서 새로고침을 하더라도 로그인 정보가 유지되도록함

anseonghyeon committed:
feat(auth): 로그인 한 유저는 로그인 버튼 클릭 시 마이페이지로 가게 라우팅함

anseonghyeon committed:
feat(auth):마이페이지에서 로그아웃하면 로그인 페이지로 라우팅되게함

anseonghyeon committed:
refactor(storage): 랭킹에 표시될 그룹들state를 저장하는 저장소를 로컬 스토리지에서 세션 스토리지로 변경함

anseonghyeon committed:
feat(order): 주문하기 페이지 구현

anseonghyeon commented:
ㅠㅠ

anseonghyeon committed:
feat(order): itemCount갯수에 따라 최종 가격이 달라지도록 변경

anseonghyeon committed:
chore(format): Prettier로 코드 포맷팅함

greenblues1190 commented on src/components/FriendSelector.tsx:
```diff
@@ -45,12 +45,14 @@ const FriendSelectorBoxText = styled.p`
 `;
 
 function FriendSelector() {
+  const userId = sessionStorage.getItem('userId')?.split('@')[0] ?? '';
```
만약 서비스 전반에서 userId와 같이 인증 정보를 사용하는 일이 많아진다면, 매번 컴포넌트에서 세션 스토리지에 접근하는 코드를 작성해야 할 것 같은데요, 보다 나은 개발 경험과 유지보수성 + 가독성까지 챙길 수 있도록 인증 정보를 관리하고 반환해주는 커스텀 훅을 만들면 좋을 것 같습니다!

`userId`라는 리액트 state를 반환하는 `useUser` 커스텀 훅을 구현해볼 수 있을까요?

```suggestion
  const { userId } = useUser();
```

greenblues1190 commented on src/components/NavBar.tsx:
```diff
@@ -61,11 +61,20 @@ const NavBarLoginBtn = styled.button`
 
 function NavBar() {
   const navigate = useNavigate();
+  const userId = sessionStorage.getItem('userId')?.split('@')[0] ?? '';
```
```suggestion
  const { userId } = useUser();
```

greenblues1190 commented on src/components/NavBar.tsx:
```diff
@@ -61,11 +61,20 @@ const NavBarLoginBtn = styled.button`
+    } else {
+      navigate('/login');
+    }
+  }
```
실무에서는 인증인가(사용자가 로그인을 하였는지, 만약 했다면 해당 기능에 접근할 권한이 있는지) 여부를 검증하고, 인증인가가 필요한 경로인지 확인하여 로그인 페이지로 리다이렉션하는 요구사항을 자동으로 처리해주는 컴포넌트를 페이지에 씌우는 패턴을 주로 사용해요.

각 경로에 대해 인증인가가 필요한지 여부를 페이지 밖에서 중앙 관리할 수 있고 각 페이지마다 중복 로직을 작성하지 않아도 돼요.

제가 찾아보니 아래 글이 가장 간단하기 인증인가를 처리하는 라우트 컴포넌트를 구현한 것 같아서 이 글 한번 따라해보시면서 코드 수정해보시면 학습에 도움이 될 것 같아요!!

https://memo-code.tistory.com/63

greenblues1190 commented on src/components/RealtimeGiftRank.tsx:
```diff
@@ -157,13 +160,51 @@ const RealtimeItemPriceTitle = styled.p`
+        JSON.stringify({ brandInfo, id, imageURL, name, price }),
+      );
+
+      navigate('/Order');
```
상품 페이지로 이동할 때, 세션스토리지에 상품 정보를 담아 이동시키면, 나중에 url을 공유했을 때 공유받은 사람이 상품페이지에 접근할 수 없는 참사가 일어납니다!

url 자체에 상태를 담는다고 생각하고 `/order/1` `/order/2`처럼 url에서 orderId를 파싱해 상품을 렌더링하는 방법으로 수정해보세요.

리액트 라우터에서는 이렇게 url의 특정 부분에 값을 담는 방법을 다이나믹 세그먼트라고 부르고 만약 orderId라는 값을 url에 포함하고 싶다면 `:orderId`처럼 작성해요. (예시: /order/:orderId)

아래 블로그글 보고 한번 수정해주세요~

https://yerim0129.tistory.com/61

greenblues1190 commented on src/pages/Login.tsx:
```diff
@@ -91,54 +91,71 @@ const validatePassword = (v: string) => {
+    // sessionStorage.setItem('userPw',`${userInfo?.pw}`);
+
     navigate(from, { replace: true });
   };
```
이 일련의 로그인 과정도 useUser라는 커스텀 훅에서 추상화해주고, 컴포넌트에서는 추상화된 함수를 꺼내쓰면 좋을 것 같아요.

```suggestion
  const { user, login } = useUser();

  const handleLogin = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    if (!isFormValid) return;

    login({ id: id.value, pw: pw.value });

    navigate(from, { replace: true });
  };
```

greenblues1190 commented on src/pages/My.tsx:
```diff
@@ -0,0 +1,32 @@
+
+function My() {
+  const userId = sessionStorage.getItem('userId')?.split('@')[0] ?? '';
+  const userEmail = sessionStorage.getItem('userId') ?? '';
```
이런 유저정보도 useUser에서 꺼내쓰도록 구현할 수 있을까요?

greenblues1190 commented on src/pages/My.tsx:
```diff
@@ -0,0 +1,32 @@
+  const navigate = useNavigate();
+
+  function logOut() {
+    sessionStorage.removeItem('userId');
```
이 로그아웃 로직도 useUser에서!

```suggestion
    // const { logout } = useUser();
    
    logout();
```

greenblues1190 commented on src/pages/Order.tsx:
```diff
@@ -0,0 +1,341 @@
+}
+
+function Order() {
+  const [selectedId, setSelectedId] = useState<number | null>(null);
+  const [selectedCardImg, setSelectedCardImg] = useState(orderCard[0].imageUrl);
+  const [selectedCardTxt, setSelectedCardTxt] = useState(
+    orderCard[0].defaultTextMessage,
+  );
+
+  const [senderName, setSenderName] = useState('');
+
+  const [receiverName, setReceiverName] = useState('');
+  const [receiverPhoneNum, setReceiverPhoneNum] = useState('');
+  const [itemCount, setItemCount] = useState(1);
```
상품 주문 폼에서의 입력값들과 검증 로직이 이리저리 흩어져 있는데 useOrderForm이라는 커스텀 훅으로 관리되면 좋을 것 같아요.

일단 아래처럼 간단하게 리팩터링부터 시작해볼까요?

```suggestion
  const {
    selectedId,
    senderName,
    receiverName,
    receiverPhoneNum,
    itemCount,
    handleChangeSelectedId,
    handleChangeSenderName,
    handleChangeReceiverName,
    handleChangeReceiverPhoneNum,
    handleChangeItemCount,
  } = useOrderForm();
```

anseonghyeon replied:
혹시 이 부분은 step3 진행하면서 리팩터링 해도될까요...?
제 능력부족, 시간관리를 잘 못해서...

greenblues1190 commented on src/pages/Order.tsx:
```diff
@@ -0,0 +1,341 @@
+  setSelectedId,
+}: OrderCardProps) {
+  function onCardClick(id: number, img: string, txt: string) {
+    setSelectedId(id);
+    setSelectedCardImg(img);
+    setSelectedCardTxt(txt);
```
selectedId만으로도 `orderCard.find(c => c.id === selectedId)`를 통해 이미지와 텍스트를 가져올 수 있어요.
카드를 선택할 때마다 id와 더불어 이미지와 텍스트를 setState해주면 상태 동기화 로직이 늘어나 복잡해져요.
이렇게 특정 상태에따라 자동으로 계산되는 상태값들을 파생 상태라고 해요.

`selectedId`만 리액트 useState로 관리하고, img와 txt는 orderCard에서 id로 찾아 꺼내 쓰는 방식으로 리팩터링할 수 있을까요?

anseonghyeon replied:
Order.tsx 파일의 수정사항은 step3하면서 수정해도될까요?

greenblues1190 reviewed (changes requested):
안녕하세요 성현님! 이번 단계는 요구사항이 많아서 쉽지 않으셨을 텐데 너무 잘 진행해주셨어요 😀 

이번 과제는 실무에서 자주 마주하는 인증/인가 처리와 폼 처리를 연습해볼 수 있는 기회인 것 같아요.
도전하며 발견할 수 있는 성장 포인트가 많으니 코멘트 하나씩 천천히 살펴보시고 적용해 보시면 좋겠습니다.

제가 코멘트를 조금 많이 남겼는데요 😏, 하나씩 살펴보시고 해보시다가 막히는 부분이 있으시면 혼자 고민하기 보다는 편하게 댓글이나 DM으로 질문 혹은 도움 요청주세요. 함께 고민하고 같이 해결할 수 있는 부분들 찾아보도록 할게요.

2단계 구현하시느라 수고많으셨어요~

anseonghyeon committed:
refactor(theme): 참조연산자로만 디자인 토큰들을 참조하는것이 편한것같아서 수정하였고 디자인 토큰을 사용하는 모든 부분을 수정하였습니다

anseonghyeon committed:
refactor(structure): 전체적인 프로젝트 구조를 개선하고 중복된 부분을 제거하였습니다

ThemeProvider와 GlobalStyle을 App.tsx로 이동, main.tsx는 엔트리 포인트로만 사용하고 라우팅과 기타 전역 스타일 설정은 App.tsx로 위임, 그리고 Home.tsx 파일을 만들어 홈페이지로 사용함

anseonghyeon committed:
chore(code): 불필요한 주석 제거로 코드 정리

anseonghyeon committed:
refactor(componet): NavBar 컴포넌트의 디자인 토큰 사용부를 수정하고 이벤트 핸들러 네이밍 개선

anseonghyeon committed:
refactor(friend-selector): FriendSelector 컴포넌트의 디자인 토큰값 변경과 네이밍을 개선

anseonghyeon committed:
style(gift-category-selector): 컴포넌트 css 값 수정

anseonghyeon committed:
refactor(realtime-gift-rank): 중복된 코드와 말도안되는 구조 개선 및 스타일 수정

anseonghyeon committed:
refactor(not-found): not found 페이지에 theme 프로바이더나 글로벌 스타일지정과 같은 중복된 코드가있어서 제거함

anseonghyeon committed:
feat(login): 입력값 검증로직을 커스텀 훅으로 옴김

anseonghyeon committed:
feat(auth): useUser 커스텀 훅으로 로그인 관리

anseonghyeon committed:
feat(route): ProtectedRoute를 사용해서 페이지 접근권한을 설정함

anseonghyeon committed:
refactor(reatime-gift-rank): 오타 수정

anseonghyeon committed:
refactor(order): 중복된 글로벌 스타일 제거

anseonghyeon committed:
refactor(order): 상품 주문 입력값을 useOrderForm 커스텀 훅으로 분리

anseonghyeon committed:
feat(order): 주문 페이지 전체 UI 작성끝

anseonghyeon committed:
feat(order): 주문하기 페이지 UI와 기능 구현완료 최종

anseonghyeon committed:
fix(order): 다양한 화면 크기에서 주문하기 버튼이 화면 맨아래에 고정되어있지 않아서 수정함

greenblues1190 reviewed (changes requested):
안녕하세요 성현님 리뷰가 늦었습니다 🙇‍♂️ 머지하겠습니다

anseonghyeon commented:
아닙니다... 제가 제때제때 못했는걸요..
step3도 빨리해서 제출하겠습니다!
